packages/world-module-erc20/src/StoreConsumer.sol

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.24;
+
+import { StoreCore } from "@latticexyz/store/src/Store.sol";
+import { StoreSwitch } from "@latticexyz/store/src/StoreSwitch.sol";
+import { RESOURCE_TABLE, RESOURCE_OFFCHAIN_TABLE } from "@latticexyz/store/src/storeResourceTypes.sol";
+import { ResourceIds } from "@latticexyz/store/src/codegen/index.sol";
+import { ResourceId, ResourceIdLib } from "@latticexyz/store/src/ResourceId.sol";
+
+import { IBaseWorld } from "@latticexyz/world/src/codegen/interfaces/IBaseWorld.sol";
+import { ResourceAccess } from "@latticexyz/world/src/codegen/tables/ResourceAccess.sol";
+import { WorldResourceIdLib } from "@latticexyz/world/src/WorldResourceId.sol";
+import { WorldContextConsumerLib } from "@latticexyz/world/src/WorldContext.sol";
+
+import { Context } from "./Context.sol";
+
+abstract contract StoreConsumer {
+  function getStore() public view returns (address) {
+    return StoreSwitch.getStoreAddress();
+  }
+
+  function _encodeResourceId(bytes2 typeId, bytes16 name) internal view virtual returns (ResourceId);
+
+  function _encodeTableId(bytes16 name) internal view returns (ResourceId) {
+    return _encodeResourceId(RESOURCE_TABLE, name);
+  }
+
+  function _encodeOffchainTableId(bytes16 name) internal view returns (ResourceId) {
+    return _encodeResourceId(RESOURCE_OFFCHAIN_TABLE, name);
+  }
+}
+
+abstract contract WithStore is Context, StoreConsumer {
+  constructor(address store) {
+    StoreSwitch.setStoreAddress(store);
+
+    if (store == address(this)) {
+      StoreCore.registerInternalTables();
+    }
+  }
+
+  function _encodeResourceId(bytes2 typeId, bytes16 name) internal view virtual override returns (ResourceId) {
+    return ResourceIdLib.encode(typeId, name);
+  }
+}
+
+abstract contract WithWorld is WithStore {
+  bytes14 public immutable NAMESPACE;
+
+  error NamespaceAlreadyExists(bytes14 namespace);
+  error CallerHasNoNamespaceAccess();
```

Comment:
A pattern we use in World and Store is to prefix the error with the contract/library name, ie `WithWorld_NamespaceAlreadyExists` - makes it a little easier to know where the error originated from in revert messages.

https://github.com/latticexyz/mud/blob/3a80bed31b97d439025f68b8e4ded27354e102f1/packages/world/src/IWorldErrors.sol#L50

Reply:
also we tend to lift this out into an interface and extend the interface when we need to ensure the errors propagate out to the ABI

for example, errors defined libraries where we make a library call in a contract won't get the errors from the library in the ABI

I _think_ this is fixed in newer solc versions but haven't checked in a while